Add googlepubsub-connector-example (#1643)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/google-pubsub-connector/googlepubsub-connector-example.md

# Google Pub Sub Connector Example  

> The Google Pub/Sub connector allows you to access the [Google Cloud Pub/Sub API Version v1](https://cloud.google.com/pubsub/docs/reference/rest/) through WSO2 EI. Google Cloud Pub/Sub is a fully-managed real-time messaging service that allows you to send and receive messages between independent applications.

Reviewer: Hence the Google Pub Sub Connector allows developers to make asynchronous messaging flows inside WSO2 EI mediation. It facilitates to achieve use-cases as below

1. One-to-many messaging. WSO2 EI can place a message in a topic and many other parties can consume it. 
2. Distributing event notifications - WSO2 EI can send events to Google Pub Sub and interested event listeners will get triggered. 
3. Streaming sensor data to Google cloud using WSO2 EI
4. Reliability improvement in processing messages. Messages received by WSO2 EI can be sent to Google Pub Sub and later received and processed by an WSO2 EI in a different region.  

Inspired from: https://cloud.google.com/pubsub/docs/overview

Author: Added.

## What you'll build

Given below is a sample scenario that demonstrates how to work with the WSO2 EI Google Pub Sub Connector to:

> 1. Create a Topic for store company update notifications. Once the response from the Google cloud is received, trigger subscription operation using the extracted information from the response.

Reviewer: for >> to

Reviewer: `Once the response from the Google cloud is received, trigger subscription operation using the extracted information from the response` this is done later, isn't it?

2. Insert company update notifications to the created topic.
3. Retrieve company updates from the created topic.

> To work with the Google Pub/Sub connector, you need to have a Google Cloud Platform account. Please refer the [Setting up the Google Pub Sub Environment](googlepubsub-connector-configuration.md) documentation. 

Reviewer: documentation to setup an account.

Author: Updated.

# Google Pub Sub Connector Example  

The Google Pub/Sub connector allows you to access the [Google Cloud Pub/Sub API Version v1](https://cloud.google.com/pubsub/docs/reference/rest/) through WSO2 EI. Google Cloud Pub/Sub is a fully-managed real-time messaging service that allows you to send and receive messages between independent applications.

> Hence the Google Pub Sub Connector allows developers to make asynchronous messaging flows inside WSO2 EI mediation. It facilitates to achieve use-cases as below

Reviewer 2: ```suggestion
The Google Pub Sub Connector allows developers to make asynchronous messaging flows inside WSO2 EI mediation. It facilitates to achieve the following use-cases.
```

Author: Updated.

1. One-to-many messaging. WSO2 EI can place a message in a topic and many other parties can consume it.
2. Distributing event notifications - WSO2 EI can send events to Google Pub Sub and interested event listeners will get triggered.
3. Streaming sensor data to Google cloud using WSO2 EI
4. Reliability improvement in processing messages. Messages received by WSO2 EI can be sent to Google Pub Sub and later received and processed by an WSO2 EI in a different region.

Inspired from: [Pubsub docs](https://cloud.google.com/pubsub/docs/overview)

## What you'll build

Given below is a sample scenario that demonstrates how to work with the WSO2 EI Google Pub Sub Connector to:

> 1. Create a Topic for store company update notifications.

Reviewer 2: ```suggestion
1. Create a Topic to store company update notifications.
```

Author: Updated.

2. Insert company update notifications to the created topic.
3. Retrieve company updates from the created topic.

To work with the Google Pub/Sub connector, you need to have a Google Cloud Platform account. Please refer the [Setting up the Google Pub Sub Environment](googlepubsub-connector-configuration.md) documentation to setup an account.

> In this scenario the user needs to create a **Topic** in **Google Cloud Platform account**, topic under **Big Data**. In this topic used to store notifications related to the company updates. Once the user invoking the `createTopic` resource, the subscribing operation also triggered simultaneously. Then the user can insert company update notifications to the created topic. Finally the user can retrieve the company updates from the subscribed topic while invoking the API.

Reviewer 2: ```suggestion
In this scenario the user needs to create a **Topic** in **Google Cloud Platform account** under **Big Data**. This topic is used to store notifications related to the company updates. Once the user invokes the `createTopic` resource, the subscribing operation also gets triggered simultaneously. Then the user can insert company update notifications to the created topic. Finally the user can retrieve the company updates from the subscribed topic while invoking the API.
```

Author: Updated.

All three operations are exposed via an API. The API with the context `/resources` has six resources.

> * `/createTopic` : Used to Create a Topic for store company notifications and subscribe to the topic. 

Reviewer 2: ```suggestion
* `/createTopic` : Used to create a Topic for store company notifications and subscribe to the topic. 
```

Author: Updated.

   ```   
   [mediation]
   synapse.enable_xpath_dom_failover="true"
   ```
   
The following diagram shows the overall solution. The user creates a topic, stores some company update notifications, and then receives it back. To invoke each operation, the user uses the same API.

<img src="../../../../assets/img/connectors/Google-pubsub-Connector1.png" title="pub-sub connector example" width="700" alt="pub-sub connector example"/>

If you do not want to configure this yourself, you can simply [get the project](#get-the-project) and run it.

## Configure the connector in WSO2 Integration Studio

Connectors can be added to integration flows in [WSO2 Integration Studio](https://wso2.com/integration/integration-studio/), which is the tooling component of WSO2 EI. Once added, the operations of the connector can be dragged onto your canvas and added to your resources.

### Import the connector

> Follow these steps to set up the ESB Solution Project and the Connector Exporter Project. 

Reviewer 2: ```suggestion
Follow these steps to set up the Integration Project and the Connector Exporter Project. 
```

Author: Updated.

        - **apiUrl** : The application URL of Google Pub/Sub.
        - **apiVersion** : The version of the Google Pub/Sub API.     
    
        <img src="../../../../assets/img/connectors/pubsub-api-init-operation.png" title="Add values to the init operation" width="800" alt="Add values to the init operation"/>

2. Set up the **createTopic** operation.

    1. Navigate into the **Palette** pane and select the graphical operations icons listed under **Googlepubsub Connector** section. Then drag and drop the `createTopic` operation into the Design pane.
           
    2. The createTopic operation creates a new topic with the name that you specify.
              
        - **projectId** : The unique ID of the project within which you want to create a topic.
        - **topicName** : The name that you want to give the topic that you are creating.
        
        While invoking the API, topicName values is populated as an input value for the operation.
        
        <img src="../../../../assets/img/connectors/pubsub-drag-and-drop.createTopic.png" title="Drag and drop createTopic operation" width="500" alt="Drag and drop createTopic operation"/>    
    

>     3. To get the input values in to the API we can use the [property mediator](../../../mediators/property-Mediator). Navigate into the **Palette** pane and select the graphical mediators icons listed under **Mediators** section. Then drag and drop the `Property` mediators into the Design pane as shown bellow.

Reviewer 2: ```suggestion
    3. To get the input values in to the API we can use the [property mediator](../../../mediators/property-Mediator). Navigate into the **Palette** pane and select the graphical mediators icons listed under **Mediators** section. Then drag and drop the `Property` mediators into the Design pane as shown below.
```

Author: Updated.

    
        <img src="../../../../assets/img/connectors/pubsub-api-drag-and-drop-property-mediator.png" title="Add property mediators" width="800" alt="Add property mediators"/>

        The parameters available for configuring the Property mediator are as follows:
    

>         > **Note**: That the properties should be add to the pallet before create the operation.

Reviewer 2: ```suggestion
        > **Note**: The properties should be added to the pallet before creating the operation.
```

Author: Updated.

    
         - **projectId** : The unique ID of the project within which the topic is created.
         - **subscriptionName** : The name of the subscription.
         - **topicName** : The name of the topic for which you want to create a subscription.
         - **ackDeadlineSeconds** :  The maximum time a subscriber can take to acknowledge a message that is received.
            
         <img src="../../../../assets/img/connectors/pubsub-api-createTopicSubscription-operation.png" title="Add values to the createTopicSubscription operation" width="800" alt="Add values to the createTopicSubscription operation"/>
    
    3. Add the property mediator to capture the `subscriptionName` values. This contains the name of the subscription.
   
        - **name** : subscriptionName
        - **expression** : json-eval($.subscriptionName)
     
        <img src="../../../../assets/img/connectors/pubsub-api-property-mediator-property2-value2.png" title="Add values to capture subscriptionName" width="600" alt="Add values to capture subscriptionName"/>  
    

>     4. Add the property mediator to sore the name of the created Topic value from the response of the createTopic operation. 

Reviewer 2: ```suggestion
    4. Add the property mediator to store the name of the created Topic value from the response of the createTopic operation. 
```

Author: Updated.

    
        - **name** : nameforsubscription
        - **expression** : json-eval($.name)
        
        <img src="../../../../assets/img/connectors/pubsub-api-property-mediator-nameforsubscription.png" title="Add values to capture nameforsubscription" width="600" alt="Add values to capture nameforsubscription"/>
        

>     5. Add the property mediator to capture Topic name from the response using the splitting separators in the results. 

Reviewer 2: ```suggestion
    5. Add the property mediator to capture the Topic name from the response using the splitting separators in the results. 
```

Author: Updated.

        
         - **name** : nameforsubscription
         - **expression** : json-eval($.name)
            
        <img src="../../../../assets/img/connectors/pubsub-api-property-mediator-splitting.png" title="Add values to capture splitting value" width="600" alt="Add values to capture splitting value"/>    

4. Forward the backend response to the API caller.
    

>     When you are invoking the created resource, the request of the message is going through the `/createTopic` resource. Finally, it is passed to the [Respond mediator](). The Respond Mediator stops the processing on the current message and sends the message back to the client as a response.            

Reviewer 2: Please add the missing relative link.

Author: Updated.

>     1. Drag and drop **respond mediator** to the **Design view**. 

Reviewer 2: ```suggestion
    1. Drag and drop the **respond mediator** to the **Design view**. 
```

Author: Updated.

       
       - **name** : data
       - **expression** : json-eval($.data)
       
       <img src="../../../../assets/img/connectors/pubsub-data.png" title="Add values to the data operation" width="800" alt="Add values to the data operation"/>

**Configure a resource for the pullMessage operation**

1. Initialize the connector.
   
   1. Initialize the connector. You can use the same configuration to initialize the connector. Please follow the steps given in section 1 for setting up the init operation to the createTopic operation
    
   2. Set up the `publishMessage` operation. Navigate into the **Palette** pane and select the graphical operations icons listed under **Googlepubsub Connector** section. Then drag and drop the `publishMessage` operation into the **Design pane**.
       
       - **projectId** : The unique ID of the project within which the topic is created.
       - **subscriptionName** : The name of the topic for which you want to create a subscription.
       - **maxMessages** :  The maximum number of messages to retrieve.
       - **returnImmediately** : Set this to true if you want the server to respond immediately.
            

>        <img src="../../../../assets/img/connectors/pubsub-pullmessages.png" title="Add values to the pull messages operation" width="800" alt="Add values to the pull messages   operation"/>

Reviewer 2: ```suggestion
       <img src="../../../../assets/img/connectors/pubsub-pullmessages.png" title="Add values to the pull messages operation" width="800" alt="Add values to the pull messages operation"/>
```

Author: Updated.

>    3. Add the property mediator to capture the `subscriptionName` values. Please follow the steps given in createTopicSubscription operation.     

Reviewer 2: ```suggestion
   3. Add the property mediator to capture the `subscriptionName` values. Follow the steps given in createTopicSubscription operation.     
```

Author: Updated.

                    </googlepubsub.init>
                    <googlepubsub.pullMessage>
                        <projectId>ei-connector-improvement</projectId>
                        <subscriptionName>{$ctx:subscriptionName}</subscriptionName>
                        <maxMessages>2</maxMessages>
                        <returnImmediately>false</returnImmediately>
                    </googlepubsub.pullMessage>
                    <respond/>
                </inSequence>
                <outSequence/>
                <faultSequence/>
            </resource>
        </api>
        ```
## Get the project

You can download the ZIP file and extract the contents to get the project code.

<a href="../../../../assets/attach/connectors/googlepubsub-connector.zip">
    <img src="../../../../assets/img/connectors/download-zip.png" width="200" alt="Download ZIP">
</a>

!!! tip
    You may need to update the simulator details and make other such changes before deploying and running this project.

## Deployment

Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 

{!references/connectors/deploy-capp.md!}   

## Testing

> Invoke the API as shown below using the curl command. Curl Application can be downloaded from [here] (https://curl.haxx.se/download.html).

Reviewer 2: ```suggestion
Invoke the API as shown below using the curl command. Curl Application can be downloaded from [here](https://curl.haxx.se/download.html).
```

Author: Updated.
